File: frontend/src/types/battery.ts
export interface TestCase {
  test_id: string;
  category: string;
  name: string;
  prompt: string;
  expected_behavior: string;
  expected_tokens?: string[];
  unexpected_tokens?: string[];
  compare_prompts?: string[];
}

export interface TestPrediction {
  token: string;
  prob: number;
}

export interface ActivationSummary {
  peak_layers: number[];
  peak_activation: number;
  active_layer_count: number;
}

export interface CompareResult {
  prompt: string;
  top_k: TestPrediction[];
  pronoun_probs?: Record<string, number>;
}

export interface SAEFeatureBrief {
  feature_idx: number;
  activation: number;
  activation_normalized: number;
  neuronpedia_url: string | null;
}

export interface TestSAEResult {
  layer_idx: number;
  hook_name: string;
  top_features: SAEFeatureBrief[];
}

export interface CrossTestFeature {
  feature_idx: number;
  neuronpedia_url: string | null;
  test_ids: string[];
  categories: string[];
  count: number;
  avg_activation: number;
  max_activation: number;
}

export interface BatterySAESummary {
  layer_idx: number;
  d_sae: number;
  total_unique_features: number;
  cross_test_features: CrossTestFeature[];
  per_category_top_features: Record<string, number[]>;
  interpretation: string;
}

export interface TestResult {
  test_id: string;
  category: string;
  name: string;
  prompt: string;
  passed: boolean;
  top_k: TestPrediction[];
  actual_token: string;
  actual_prob: number;
  expected_prob?: number;
  activation_summary: ActivationSummary;
  interpretation: string;
  compare_results?: CompareResult[];
  sae_features?: TestSAEResult;
}

export interface BatteryResult {
  model_id: string;
  total_tests: number;
  passed: number;
  failed: number;
  results: TestResult[];
  summary: string;
  sae_summary?: BatterySAESummary;
}

File: backend/neural_mri/schemas/battery.py
from __future__ import annotations

from pydantic import BaseModel


class TestCase(BaseModel):
    test_id: str
    category: str
    name: str
    prompt: str
    expected_behavior: str
    expected_tokens: list[str] | None = None
    unexpected_tokens: list[str] | None = None
    compare_prompts: list[str] | None = None


class TestPrediction(BaseModel):
    token: str
    prob: float


class ActivationSummary(BaseModel):
    peak_layers: list[int]
    peak_activation: float
    active_layer_count: int


class CompareResult(BaseModel):
    prompt: str
    top_k: list[TestPrediction]
    pronoun_probs: dict[str, float] | None = None  # e.g. {" he": 0.3, " she": 0.1}


class TestResult(BaseModel):
    test_id: str
    category: str
    name: str
    prompt: str
    passed: bool
    top_k: list[TestPrediction]
    actual_token: str
    actual_prob: float
    expected_prob: float | None = None
    activation_summary: ActivationSummary
    interpretation: str
    compare_results: list[CompareResult] | None = None


class BatteryResult(BaseModel):
    model_id: str
    total_tests: int
    passed: int
    failed: int
    results: list[TestResult]
    summary: str


class BatteryRunRequest(BaseModel):
    categories: list[str] | None = None  # None = all categories
    locale: str = "en"
